File: src/GUI/objects/Pickers/Picker.java

package gui.objects.Pickers;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Stream;

import javafx.event.Event;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.*;

import util.animation.Anim;
import util.functional.Functors.Ƒ1;
import util.parsing.ToStringConverter;

import static java.lang.Math.*;
import static java.util.Arrays.asList;
import static javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER;
import static javafx.scene.input.MouseButton.PRIMARY;
import static javafx.scene.input.MouseButton.SECONDARY;
import static javafx.scene.input.MouseEvent.MOUSE_DRAGGED;
import static javafx.scene.input.MouseEvent.MOUSE_PRESSED;
import static util.animation.Anim.seq;
import static util.functional.Util.*;

/**
 * Generic item picker.
 * <p>
 * Node displaying elements as grid.
 * Elements are converted to their text representation according to provided
 * mapper. Element should override toString() method if no mapper is provided.
 * <p>
 * Elements will be sorted lexicographically.
 * <p>
 * Elements will be represented graphically depending on the cell factory.
 * <p>
 * @author Plutonium_
 */
public class Picker<E> {

    /** Style class for cell. */
    public static final String STYLE_CLASS = "item-picker";
    public static final List<String> CELL_STYLE_CLASS = asList("block","item-picker-element");
    /** Default on select action. Does nothing. */
    public static final Consumer DEF_onSelect = item -> {};
    /** Default on cancel action. Does nothing. */
    public static final Runnable DEF_onCancel = () -> {};
    /** Default Text factory. Uses item's toString() method. */
    public static final ToStringConverter DEF_textCoverter = Object::toString;
    /** Default Item supplier. Returns empty stream. */
    public static final Supplier<Stream> DEF_itemSupply = Stream::empty;

    public boolean consumeCancelClick = false;

    private final CellPane tiles = new CellPane();
    public final ScrollPane root = new ScrollPane(tiles);

    /**
     * Procedure executed when item is selected passing the item as parameter.
     * Default implementation does nothing. Must not be null;
     */
    public Consumer<E> onSelect = DEF_onSelect;
    /**
     * Procedure executed when no item is selected. Invoked when user cancels
     * the picking by right click.
     * Default implementation does nothing. Must not be null;
     * <p>
     * For example one might want to close this picker when no item is selected.
     */
    public Runnable onCancel = DEF_onCancel;
    /**
     * Text factory.
     * Creates string representation of the item.
     * Default implementation uses item's toString() method.
     * Must not be null.
     */
    public ToStringConverter<E> textCoverter = DEF_textCoverter;
    /**
     * Item supplier. Fetches the items as a stream.
     * Default implementation returns empty stream. Must not be null;
     */
    public Supplier<Stream<E>> itemSupply = (Supplier)DEF_itemSupply;
    /**
     * Cell factory.
     * Creates graphic representation of the item.
     * Also might define minimum and maximum item size.
     * Must not be null;
     */
    public Ƒ1<E,Region> cellFactory = item -> {
        String text = textCoverter.toS(item);
        Label l = new Label(text);
        StackPane b = new StackPane(l);
        b.getStyleClass().setAll(CELL_STYLE_CLASS);
        StackPane a = new StackPane(b);
        a.setMinSize(90, 30);
        return a;
    };

    public Picker() {
        root.setPannable(false);  // forbid mouse panning
        root.setHbarPolicy(NEVER);
        root.setPrefSize(-1,-1);  // leave resizable
        root.setFitToWidth(true); // make content resize with scroll pane
        // consume problematic events and prevent from propagating
        // disables unwanted behavior of the popup
        root.addEventFilter(MOUSE_PRESSED, Event::consume);
        root.addEventFilter(MOUSE_DRAGGED, Event::consume);
        root.setOnMouseClicked(e->{
            if(e.getButton()==SECONDARY) {
                onCancel.run();
                if(consumeCancelClick) e.consume();
            }
            e.consume();
        });
        root.getStyleClass().add(STYLE_CLASS);
    }

    public void buildContent() {
        tiles.getChildren().clear();
        // get items
        itemSupply.get()
            // & sort
            .sorted(byNC(textCoverter::toS))
            // & create cells
            .forEach( item -> {
                Node cell = cellFactory.apply(item);
                     cell.setOnMouseClicked( e -> {
                         if(e.getButton()==PRIMARY) {
                            onSelect.accept(item);
                            e.consume();
                         }
                     });
                tiles.getChildren().add(cell);
            });

        // animate & show
        int s = getCells().size();
        Anim.par(getCells(), (i,n) -> seq(
            new Anim(n::setOpacity).dur(i*(750/s)).intpl(0),
            new Anim(n::setOpacity).dur(500).intpl(x -> sqrt(x))
        )).play();
    }

    public Node getNode() {
        buildContent();
        return root;
    }

    public List<Region> getCells() {
        return (List) list(tiles.getChildren());
    }

    private class CellPane extends Pane {

        @Override
        protected void layoutChildren() {
            double width = root.getWidth();
            double height = root.getHeight();


            int gap = 5;
            int elements = tiles.getChildren().size();
            double min_cell_w = max(1,getCells().get(0).getMinWidth());
            double min_cell_h = max(1,getCells().get(0).getMinHeight());
            // if(elements==0) return;

            int c = width>height ? (int) ceil(sqrt(elements)) : (int) floor(sqrt(elements));
                c = width<c*min_cell_w ? (int)floor(width/min_cell_w) : c;
            final int columns = max(1,c);

            int rows = (int) ceil(elements/(double)columns);

            double sumgapy = (rows-1) * gap;
            final double cell_height = height<rows*min_cell_h ? min_cell_h : (height-sumgapy)/rows-1/(double)rows;

            double W = rows*(cell_height+gap)-gap>height ? width-15 : width; // take care of scrollbar
            double sumgapx = (columns-1) * gap;  // n elements have n-1 gaps
            final double cell_width = (W-sumgapx)/columns;

            forEachWithI(getCells(), (i,n) -> {
                double x = i%columns * (cell_width+gap);
                double y = i/columns * (cell_height+gap);
                n.relocate(x,y);
                n.resize(cell_width, cell_height);
            });
        }

    }
}

File: src/player/main/sp/it/pl/gui/objects/picker/Picker.kt
package sp.it.pl.gui.objects.picker

import javafx.geometry.Insets
import javafx.geometry.VPos.CENTER
import javafx.scene.control.ScrollPane
import javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseButton.SECONDARY
import javafx.scene.input.MouseEvent.MOUSE_CLICKED
import javafx.scene.layout.Pane
import javafx.scene.layout.Region
import javafx.scene.layout.Region.USE_COMPUTED_SIZE
import javafx.scene.text.TextAlignment
import sp.it.util.animation.Anim.Companion.anim
import sp.it.util.collections.setTo
import sp.it.util.functional.supplyIf
import sp.it.util.math.max
import sp.it.util.reactive.attach
import sp.it.util.reactive.onEventDown
import sp.it.util.ui.label
import sp.it.util.ui.lay
import sp.it.util.ui.pane
import sp.it.util.ui.pseudoClassChanged
import sp.it.util.ui.scrollTextCenter
import sp.it.util.ui.setMinPrefMaxSize
import sp.it.util.ui.setScaleXY
import sp.it.util.ui.stackPane
import sp.it.util.ui.text
import sp.it.util.units.millis
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Generic picker.
 *
 * The elements are obtained using [itemSupply] when [buildContent] is called, sorted lexicographically and displayed
 * in a 2D grid using [textConverter], [infoConverter].
 */
open class Picker<E> {

    private val tiles = CellPane()
    /** Scene graph root of this object. */
    val root = ScrollPane(tiles)
    /** Invoked when item is selected. Default implementation does nothing. */
    var onSelect: (E) -> Unit = {}
    /** Invoked when user user cancels the picking. Default implementation does nothing. */
    var onCancel: () -> Unit = {}
    /** It may be desirable to consume the mouse click event that caused the cancellation. Default false. */
    var consumeCancelEvent = false
    /** Cell text factory producing name/title of the item. Default implementation calls [Any.toString] */
    var textConverter: (E) -> String = Any?::toString
    /** Cell detail text text factory producing description of the item. Default implementation returns empty string. */
    var infoConverter: (E) -> String = { "" }
    /** Supplier that returns items to be displayed. Default implementation returns empty sequence. */
    var itemSupply: () -> Sequence<E> = { sequenceOf() }

    private val cellFactory: (E) -> Pane = { item ->
        stackPane {
            setMinSize(90.0, 30.0)
            styleClass += CELL_STYLE_CLASS
            padding = Insets(20.0)

            val contentText = textConverter(item)
            val contentInfoText = infoConverter(item)
            val content = label(contentText)

            lay += content
            lay += supplyIf(contentInfoText.isNotEmpty()) {
                val contentInfo = scrollTextCenter {
                    text(contentInfoText) {
                        isMouseTransparent = true
                        textOrigin = CENTER
                        textAlignment = TextAlignment.CENTER
                    }
                }

                val anim = anim(300.millis) {
                    content.opacity = 1-it*it
                    contentInfo.opacity = it
                    contentInfo.setScaleXY(0.7+0.3*it*it)
                }

                anim.applyNow()
                hoverProperty() attach { anim.playFromDir(it) }

                contentInfo
            }
        }
    }

    constructor() {
        root.apply {
            setMinPrefMaxSize(USE_COMPUTED_SIZE)
            isPannable = false  // forbid mouse panning (can cause unwanted horizontal scrolling)
            isFitToWidth = true // make content horizontally resize with scroll pane
            hbarPolicy = NEVER
            onEventDown(MOUSE_CLICKED, SECONDARY, false) {
                onCancel()
                if (consumeCancelEvent) it.consume()
            }
            styleClass += STYLE_CLASS
        }
    }

    fun buildContent() {
        tiles.children setTo itemSupply()
                .sortedBy(textConverter)
                .map { item ->
                    cellFactory(item).apply {
                        onEventDown(MOUSE_CLICKED, PRIMARY) { onSelect(item) }
                    }
                }
        tiles.children += pane {
            styleClass += CELL_STYLE_CLASS
            pseudoClassChanged("empty", true)
            properties[KEY_EMPTY_CELL] = null
            isManaged = false
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun getCells() = tiles.children.filter { !it.properties.containsKey(KEY_EMPTY_CELL) } as List<Region>

    private inner class CellPane: Pane() {

        override fun layoutChildren() {
            val cells = getCells()
            val padding = root.padding
            val gap = 1.0
            val width = root.width-padding.left-padding.right
            val height = root.height-padding.top-padding.bottom

            if (cells.isEmpty()) return

            val elements = cells.size
            val cellMinWidth = 1.0 max cells.first().minWidth
            val cellMinHeight = 1.0 max cells.first().minHeight

            var c = if (width>height) ceil(sqrt(elements.toDouble())).toInt() else floor(sqrt(elements.toDouble())).toInt()
            c = if (width<c*cellMinWidth) floor(width/cellMinWidth).toInt() else c
            val columns = max(1, c)

            val rows = ceil(elements.toDouble()/columns).toInt()

            val gapSumY = (rows-1)*gap
            val cellHeight = if (height<rows*cellMinHeight) cellMinHeight else (height-gapSumY)/rows-1.0/rows

            val w = if (rows*(cellHeight+gap)-gap>height) width-15 else width // TODO: take care of scrollbar better
            val gapSumX = (columns-1)*gap
            val cellWidth = (w-gapSumX)/columns

            cells.forEachIndexed { i, n ->
                val x = padding.left+i%columns*(cellWidth+gap)
                val y = padding.top+i/columns*(cellHeight+gap)
                n.resizeRelocate(
                        x.snapX,
                        y.snapY,
                        (x+cellWidth).snapX-x.snapX,
                        (y+cellHeight).snapY-y.snapY
                )
            }

            val needsEmptyCell = cells.isEmpty() || cells.size!=columns*rows
            val emptyCell = children.find { it.properties.containsKey(KEY_EMPTY_CELL) }!!
            if (needsEmptyCell) {
                val i = cells.size
                val x = padding.left+i%columns*(cellWidth+gap)
                val y = padding.top+i/columns*(cellHeight+gap)
                emptyCell.resizeRelocate(
                        x.snapX,
                        y.snapY,
                        (root.width-padding.right).snapX-x.snapX,
                        (y+cellHeight).snapY-y.snapY
                )
            } else {
                emptyCell.resizeRelocate(0.0, 0.0, 0.0, 0.0)
            }
        }

        private val Double.snapX: Double get() = this//snapPositionX(this)
        private val Double.snapY: Double get() = this//snapPositionY(this)
    }

    companion object {
        const val STYLE_CLASS = "item-picker"
        val CELL_STYLE_CLASS = listOf("block", "item-picker-element")
        private const val KEY_EMPTY_CELL = "empty_cell"
    }

}